exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinMemoryCalculatorImpl.java

```diff
@@ -387,9 +387,12 @@ private void calculateMemoryUsage()
         long incompletePartitionsBatchSizes = ((long) partitions) * partitionBuildBatchSize;
         // We need to reserve all the space for incomplete batches, and the incoming batch as well as the
         // probe batch we sniffed.
-        // TODO when batch sizing project is complete we won't have to sniff probe batches since
-        // they will have a well defined size.
-        reservedMemory = incompletePartitionsBatchSizes + maxBuildBatchSize + probeSizePredictor.getBatchSize();
+        reservedMemory = incompletePartitionsBatchSizes + maxBuildBatchSize;
+
+        if (!firstCycle) {
+          // If this is NOT the first cycle the HashJoin operator owns the probe batch and we need to reserve space for it.
+          reservedMemory += probeSizePredictor.getBatchSize();
```

Comment:
Sounds correct, but note that this method *calculateMemoryUsage()* is invoked from the *innerNext()* which is called (recursively) after both the first Build and first Probe side spilled batches are already read.
 -  In line 363 *maxProbeBatchSize* is calculated; how different is it from getBatchSize() here ?
 -  Should the build side spilled "incoming" batch size also be taken into account ?

Reply:
- This should work correctly with spilling. innerNext calls executeBuild which creates a new calculator each time. The calculator is initialized with the current build and probe batches, so the calculator will always be initialized with the correct state.
 - You are right, there should be no difference between maxProbeBatchSize and getBatchSize()  now, since this code will only be activated when reading spilled batches. This is leftover logic from when the calculator would compute the worst case incoming probe batch size during the first cycle. I will consolidate these two sizes to clean things up.
 - The build side incoming batch is taken into account on line 390. *Note:* this code still reserves memory for the incoming build side batch during the first cycle, which is not correct. I want to fix that issue in a separate PR though, since these changes get quite large and require extensive changes to the unit tests.